concordium-consensus/src/Concordium/KonsensusV1/Types.hs

```diff
@@ -0,0 +1,804 @@
+      bbTransactions :: !(Vector.Vector BlockItem),
+      -- |Hash of the transaction outcomes.
+      bbTransactionOutcomesHash :: !TransactionOutcomesHash,
+      -- |Hash of the block state.
+      bbStateHash :: !StateHash
+    }
+    deriving (Eq)
+
+-- |Flags indicating which optional values are set in a 'BakedBlock'.
+data BakedBlockFlags = BakedBlockFlags
+    { bbfTimeoutCertificate :: !Bool,
+      bbfEpochFinalizationEntry :: !Bool
+    }
+
+-- |Get the 'BakedBlockFlags' associated with a 'BakedBlock'.
+bakedBlockFlags :: BakedBlock -> BakedBlockFlags
+bakedBlockFlags BakedBlock{..} =
+    BakedBlockFlags
+        { bbfTimeoutCertificate = isPresent bbTimeoutCertificate,
+          bbfEpochFinalizationEntry = isPresent bbEpochFinalizationEntry
+        }
+
+instance Serialize BakedBlockFlags where
+    put BakedBlockFlags{..} = putWord8 bits
+      where
+        bits = toBit 0 bbfTimeoutCertificate .|. toBit 1 bbfEpochFinalizationEntry
+        toBit _ False = 0
+        toBit i True = bit i
+    get = label "BakedBlockFlags" $ do
+        bits <- getWord8
+        unless (bits == bits .&. 0b11) $ fail "Invalid BakedBlockFlags."
+        return $
+            BakedBlockFlags
+                { bbfTimeoutCertificate = testBit bits 0,
+                  bbfEpochFinalizationEntry = testBit bits 1
+                }
+
+-- |Serialize a 'BakedBlock'.
+putBakedBlock :: Putter BakedBlock
+putBakedBlock bb@BakedBlock{..} = do
+    put bbRound
+    put bbEpoch
+    put bbTimestamp
+    put bbBaker
+    put bbBakerKey
+    put bbQuorumCertificate
+    put (bakedBlockFlags bb)
+    mapM_ put bbTimeoutCertificate
+    mapM_ put bbEpochFinalizationEntry
+    put bbNonce
+    put bbStateHash
+    put bbTransactionOutcomesHash
+    putWord64be (fromIntegral (Vector.length bbTransactions))
+    mapM_ putBlockItemV0 bbTransactions
+
+-- |Deserialize a 'BakedBlock'. The protocol version is used to determine which transaction
+-- types are allowed in the block.
+getBakedBlock :: SProtocolVersion pv -> TransactionTime -> Get BakedBlock
+getBakedBlock spv tt = label "BakedBlock" $ do
+    bbRound <- get
+    bbEpoch <- get
+    bbTimestamp <- get
+    bbBaker <- get
+    bbBakerKey <- get
+    bbQuorumCertificate <- get
+    BakedBlockFlags{..} <- get
+    bbTimeoutCertificate <-
+        if bbfTimeoutCertificate
+            then Present <$> get
+            else return Absent
+    bbEpochFinalizationEntry <-
+        if bbfEpochFinalizationEntry
+            then Present <$> get
+            else return Absent
+    bbNonce <- get
+    bbStateHash <- get
+    bbTransactionOutcomesHash <- get
+    numTrans <- getWord64be
+    -- We check that there is at least one byte remaining in the serialization per transaction.
+    -- This is to prevent a malformed block from causing us to allocate an excessively large vector,
+    -- as this could lead to an out-of-memory error.
+    remBytes <- remaining
+    when (fromIntegral remBytes < numTrans) $
```

Comment:
There is a constant `transactionHeaderSize` or some such that is better suited to this and will give a better safety margin.